osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
 
-            foreach (var positionInfo in positionInfos)
+            for (int i = 0; i < positionInfos.Count; i++)
             {
-                // rateOfChangeMultiplier only changes every 5 iterations in a combo
-                // to prevent shaky-line-shaped streams
-                if (positionInfo.HitObject.IndexInCurrentCombo % 5 == 0)
-                    rateOfChangeMultiplier = (float)rng.NextDouble() * 2 - 1;
+                bool invertFlow = false;
+
+                if (i == 0 ||
+                    (positionInfos[i - 1].HitObject.NewCombo && (i <= 1 || !positionInfos[i - 2].HitObject.NewCombo) && (i <= 2 || !positionInfos[i - 3].HitObject.NewCombo)) ||
+                    OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject, true) ||
+                    (OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject) && rng.NextDouble() < 0.25))
+                {
+                    sequenceOffset = OsuHitObjectGenerationUtils.RandomGaussian(rng, 0, 0.02f);
+
+                    if (rng.NextDouble() < 0.6)
+                        invertFlow = true;
+                }
 
-                if (positionInfo == positionInfos.First())
+                if (i == 0)
                 {
-                    positionInfo.DistanceFromPrevious = (float)(rng.NextDouble() * OsuPlayfield.BASE_SIZE.Y / 2);
-                    positionInfo.RelativeAngle = (float)(rng.NextDouble() * 2 * Math.PI - Math.PI);
+                    positionInfos[i].DistanceFromPrevious = (float)(rng.NextDouble() * OsuPlayfield.BASE_SIZE.Y / 2);
+                    positionInfos[i].RelativeAngle = (float)(rng.NextDouble() * 2 * Math.PI - Math.PI);
                 }
                 else
                 {
-                    positionInfo.RelativeAngle = rateOfChangeMultiplier * 2 * (float)Math.PI * Math.Min(1f, positionInfo.DistanceFromPrevious / (playfield_diagonal * 0.5f));
+                    float flowChangeOffset = 0;
+                    float oneTimeOffset = OsuHitObjectGenerationUtils.RandomGaussian(rng, 0, 0.03f);
+
+                    if (positionInfos[i - 1].HitObject.NewCombo && (i <= 1 || !positionInfos[i - 2].HitObject.NewCombo) && rng.NextDouble() < 0.6)
+                    {
+                        flowChangeOffset = OsuHitObjectGenerationUtils.RandomGaussian(rng, 0, 0.05f);
+
+                        if (rng.NextDouble() < 0.8)
+                            invertFlow = true;
+                    }
+
+                    if (invertFlow)
+                        flowDirection ^= true;
+
+                    positionInfos[i].RelativeAngle = OsuHitObjectGenerationUtils.GetRelativeTargetAngle(
+                        positionInfos[i].DistanceFromPrevious,
+                        (sequenceOffset + oneTimeOffset) * (float)Math.Sqrt(positionInfos[i].DistanceFromPrevious) +
+                        flowChangeOffset * (float)Math.Sqrt(640 - positionInfos[i].DistanceFromPrevious),
```

Comment:
What is this doing?

Reply:
640 is the playfield diagonal (I re-added the field `playfield_diagonal` to make this clear). I also removed the square roots (and adjusted all the standard deviations accordingly) because it doesn't make that much of a difference here and just makes things more complex.
`sequenceOffset` and `oneTimeOffset` should affect further spaced objects more and `flowChangeOffset` should mainly affect streams, that's why they are dependent on `positionInfos[i].DistanceFromPrevious`.